designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -72,77 +72,134 @@ class DeploymentServiceImpl(
   )(implicit user: LoggedUser, ec: ExecutionContext): Future[Unit] = {
     val actionType = ProcessActionType.Cancel
 
-    validateDeploymentComment(comment).flatMap { comment =>
-      checkCanPerformActionAndAddInProgressAction[Unit](
+    for {
+      // 1. common for all commands/actions
+      // TODO: extract param definitions to do generic "parameter validation"
+      _ <- validateDeploymentComment(comment)
+      ctx <- prepareCommandContextWithAction[Unit](
         processId,
         actionType,
-        _.lastDeployedAction.map(_.processVersionId),
+        p => p.lastDeployedAction.map(_.processVersionId),
         _ => None
-      ).flatMap { case (processDetails, actionId, versionOnWhichActionIsDone, buildInfoProcessIngType) =>
-        runDeploymentActionWithNotifications(
-          actionType,
-          actionId,
-          processId,
-          versionOnWhichActionIsDone,
-          comment,
-          buildInfoProcessIngType
-        ) {
-          dispatcher
-            .deploymentManagerUnsafe(processDetails.processingType)
-            .processCommand(CancelScenarioCommand(processId.name, user.toManagerUser))
-        }.map(_ => ())
-      }
-    }
+      )
+      // 2. command specific section
+      deploymentComment = comment.map(DeploymentComment.unsafe)
+      command           = CancelScenarioCommand(processId.name, user.toManagerUser)
+      // 3. common for all commands/actions
+      _ = runActionAndHandleResults(
+        actionType,
+        ctx.actionId,
```

Comment:
You already have a nice context class, why do you pass all fields separately? :)

Reply:
Good point :) fixed.